In [1]:
import numpy as np
import pandas as pd
import matplotlib.pyplot as plt
import seaborn as sns

In [2]:
import io
import requests
url = "https://api.covidtracking.com/v1/states/daily.csv" # Access date and time : Feb 8, 2021, 10.30 A.M.
s=requests.get(url).content

In [3]:
df = pd.read_csv(io.StringIO(s.decode('utf-8')))

In [4]:
df.to_csv('../Dataset/dataset.csv', index=False, header=True) # Saving the dataset for reuse in future time

In [4]:
pd.set_option('display.max_columns', None) # To display all columns
df.head()

date state  positive  probableCases   negative  pending  \
0  20210208    AK   53694.0            NaN        NaN      NaN   
1  20210208    AL  473348.0       101577.0  1820080.0      NaN   
2  20210208    AR  307373.0        62960.0  2290742.0      NaN   
3  20210208    AS       0.0            NaN     2140.0      NaN   
4  20210208    AZ  782887.0        51100.0  2820314.0      NaN   

  totalTestResultsSource  totalTestResults  hospitalizedCurrently  \
0        totalTestsViral         1555532.0                   39.0   
1  totalTestsPeopleViral         2191851.0                 1524.0   
2        totalTestsViral         2535155.0                  777.0   
3        totalTestsViral            2140.0                    NaN   
4        totalTestsViral         6993662.0                 2853.0   

   hospitalizedCumulative  inIcuCurrently  inIcuCumulative  \
0                  1219.0             NaN              NaN   
1                 43383.0             NaN           2577.0   
2                 14099.0           274.0              NaN   
3                     NaN             NaN              NaN   
4                 54713.0           828.0              NaN   

   onVentilatorCurrently  onVentilatorCumulative  recovered dataQualityGrade  \
0                    9.0                     NaN        NaN                A   
1                    NaN                  1460.0   252880.0                A   
2                  142.0                  1458.0   286917.0               A+   
3                    NaN                     NaN        NaN              NaN   
4                  536.0                     NaN   108495.0               A+   

      lastUpdateEt          dateModified  checkTimeEt    death  hospitalized  \
0   2/8/2021 03:59  2021-02-08T03:59:00Z  02/07 22:59    279.0        1219.0   
1   2/8/2021 11:00  2021-02-08T11:00:00Z  02/08 06:00   8523.0       43383.0   
2   2/8/2021 00:00  2021-02-08T00:00:00Z  02/07 19:00   5106.0       14099.0   
3  12/1/2020 00:00  2020-12-01T00:00:00Z  11/30 19:00      0.0           NaN   
4   2/8/2021 00:00  2021-02-08T00:00:00Z  02/07 19:00  14055.0       54713.0   

            dateChecked  totalTestsViral  positiveTestsViral  \
0  2021-02-08T03:59:00Z        1555532.0             64862.0   
1  2021-02-08T11:00:00Z              NaN                 NaN   
2  2021-02-08T00:00:00Z        2535155.0                 NaN   
3  2020-12-01T00:00:00Z           2140.0                 NaN   
4  2021-02-08T00:00:00Z        6993662.0                 NaN   

   negativeTestsViral  positiveCasesViral  deathConfirmed  deathProbable  \
0           1488905.0                 NaN             NaN            NaN   
1                 NaN            371771.0          6753.0         1770.0   
2           2290742.0            244413.0          4081.0         1025.0   
3                 NaN                 0.0             NaN            NaN   
4                 NaN            731787.0         12487.0         1568.0   

   totalTestEncountersViral  totalTestsPeopleViral  totalTestsAntibody  \
0                       NaN                    NaN                 NaN   
1                       NaN              2191851.0                 NaN   
2                       NaN                    NaN                 NaN   
3                       NaN                    NaN                 NaN   
4                       NaN              3552101.0            423799.0   

   positiveTestsAntibody  negativeTestsAntibody  totalTestsPeopleAntibody  \
0                    NaN                    NaN                       NaN   
1                    NaN                    NaN                  109716.0   
2                    NaN                    NaN                       NaN   
3                    NaN                    NaN                       NaN   
4                    NaN                    NaN                       NaN   

   positiveTestsPeopleAntibody  negativeTestsPeopleAntibody  \
0                          NaN                 

In [5]:
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 19261 entries, 0 to 19260
Data columns (total 55 columns):
 #   Column                       Non-Null Count  Dtype  
---  ------                       --------------  -----  
 0   date                         19261 non-null  int64  
 1   state                        19261 non-null  object 
 2   positive                     19072 non-null  float64
 3   probableCases                8246 non-null   float64
 4   negative                     15334 non-null  float64
 5   pending                      2004 non-null   float64
 6   totalTestResultsSource       19261 non-null  object 
 7   totalTestResults             19159 non-null  float64
 8   hospitalizedCurrently        15907 non-null  float64
 9   hospitalizedCumulative       11865 non-null  float64
 10  inIcuCurrently               10400 non-null  float64
 11  inIcuCumulative              3468 non-null   float64
 12  onVentilatorCurrently        8258 non-null   float64
 13  onVentilatorCumu

## Metadata information

1. __checkTimeEt__<br>
_Field type:string_<br>
Deprecated<br>

2. __commercialScore__<br>
_Field type:integer_<br>
Deprecated<br>
_Returns : null, if no data is available_

3. __dataQualityGrade__<br>
Field type:string<br>
Data Quality Grade<br>
The COVID Tracking Project grade of the completeness of the data reporting by a state.<br>

4. __date__<br>
_Field type:integer_<br>
Date on which data was collected by The COVID Tracking Project.<br>

5. __dateChecked__<br>
Field type:string<br>
Deprecated<br>

6. __dateModified__<br>
_Field type:string_<br>
Deprecated, use lastUpdateEt instead<br>

7. __death__<br>
_Field type:integer_<br>
Deaths (confirmed and probable)<br>
Total fatalities with confirmed OR probable COVID-19 case diagnosis (per the expanded CSTE case definition of April 5th, 2020 approved by the CDC). In some states, these individuals must also have COVID-19 listed on the death certificate to count as a COVID-19 death. When states post multiple numbers for fatalities, the metric includes only deaths with COVID-19 listed on the death certificate, unless deaths among cases is a more reliable metric in the state.<br>
_Returns : null. if no data is available_<br>

8. __deathConfirmed__<br>
_Field type:integer_<br>
Deaths (confirmed)<br>
Total fatalities with confirmed COVID-19 case diagnosis (per the expanded CSTE case definition of April 5th, 2020 approved by the CDC). In some states, these individuals must also have COVID-19 listed on the death certificate to count as a COVID-19 death. When states post multiple numbers for confirmed fatalities, the metric includes only lab-confirmed deaths with COVID-19 listed on the death certificate, unless deaths among confirmed cases is a more reliable metric in the state.<br>
_Returns : null, if no data is available_

9. __deathIncrease__<br>
_Field type:integer_<br>
New deaths<br>
Daily increase in death, calculated from the previous day’s value.<br>
_Returns: null, if no data is available_<br>

10. __deathProbable__<br>
_Field type:integer_<br>
Deaths (probable)<br>
Total fatalities with probable COVID-19 case diagnosis (per the expanded CSTE case definition of April 5th, 2020 approved by the CDC). In some states, these individuals must also have COVID-19 listed on the death certificate to count as a COVID-19 death. When states post multiple numbers for probable fatalities, the metric includes only probable fatalities with COVID-19 listed on the death certificate, unless deaths among probable cases is a more reliable metric in the state.<br>
_Returns : null, if no data is available_<br>

11. __fips__<br>
_Field type:string_<br>
FIPS code<br>
State FIPS code<br>

12. __grade__<br>
_Field type:string_<br>
Deprecated<br>

13. __hash__<br>
_Field type:string_<br>
Deprecated - A hash of the current record.<br>

14. __hospitalized__<br>
_Field type:integer_<br>
Deprecated<br>
_Returns : null, if no data is available_

15. __hospitalizedCumulative__<br>
_Field type:integer_<br>
Cumulative hospitalized/Ever hospitalized<br>
Total number of individuals who have ever been hospitalized with COVID-19. Definitions vary by state / territory, and it is not always clear whether pediatric patients are included in this metric. Where possible, we report patients hospitalized with confirmed or suspected COVID-19 cases.<br>
_Returns : null, if no data is available_

16. __hospitalizedCurrently__<br>
_Field type:integer_<br>
Currently hospitalized/Now hospitalized<br>
Individuals who are currently hospitalized with COVID-19. Definitions vary by state / territory, and it is not always clear whether pediatric patients are included in this metric. Where possible, we report patients hospitalized with confirmed or suspected COVID-19 cases.<br>
_Returns : null, if no data is available_<br>

17. __hospitalizedIncrease__<br>
_Field type:integer_<br>
New total hospitalizations<br>
Daily increase in hospitalizedCumulative, calculated from the previous day’s value.<br>
_Returns : null, if no data is available_

18. __inIcuCumulative__<br>
_Field type:integer_<br>
Cumulative in ICU/Ever in ICU<br>
Total number of individuals who have ever been hospitalized in the Intensive Care Unit with COVID-19. Definitions vary by state / territory, and it is not always clear whether pediatric patients are included in this metric. Where possible, we report patients in the ICU with confirmed or suspected COVID-19 cases.<br>
_Returns : null, if no data is available<br>

19. __inIcuCurrently__<br>
_Field type:integer_<br>
Currently in ICU/Now in ICU<br>
Individuals who are currently hospitalized in the Intensive Care Unit with COVID-19. Definitions vary by state / territory, and it is not always clear whether pediatric patients are included in this metric. Where possible, we report patients in the ICU with confirmed or suspected COVID-19 cases.<br>
_Returns : null, if no data is available_<br>

20. __lastUpdateEt__<br>
_Field type:string_<br>
Last Update (ET)<br>
Date and time in Eastern time the state or territory last updated the data.<br>

21. __negative__<br>
_Field type:integer_<br>
Negative PCR tests (people)<br>
Total number of unique people with a completed PCR test that returns negative. For states / territories that do not report this number directly, we compute it using one of several methods, depending on which data points the state provides. Due to complex reporting procedures, this number might be mixing units and therefore, at best, it should only be considered an estimate of the number of people with a completed PCR test that return negative.<br>
_Returns : null, if no data is available_

22. __negativeIncrease__<br>
_Field type:integer_<br>
Deprecated<br>
_Returns : null, if no data is available_

23. __negativeRegularScore__<br>
_Field type:integer_<br>
Deprecated<br>
_Returns : null, if no data is available_<br>

24. __negativeScore__<br>
_Field type:integer_<br>
Deprecated<br>
_Returns : null, if no data is available_<br>

25. __negativeTestsAntibody__<br>
_Field type:integer_<br>
Negative antibody tests (specimens)<br>
The total number of completed antibody tests that return negative as reported by the state or territory.<br>
_Returns : null, if no data is available_<br>

26. __negativeTestsPeopleAntibody__<br>
_Field type:integer_<br>
Negative antibody tests (people)<br>
The total number of unique people with completed antibody tests that return negative as reported by the state or territory.<br>
_Returns : null, if no data is available_<br>

27. __negativeTestsViral__<br>
_Field type:integer_<br>
Negative PCR tests (specimens)<br>
Total number of completed PCR tests (or specimens tested) that return negative as reported by the state or territory. For states/territories that do not report this number directly, we compute it using one of several methods, depending on which data points the state provides. If we discover that a jurisdiction is including antigen tests in this metric, we will annotate that state or territory’s data accordingly.<br>
_Returns : null, if no data is available_<br>

28. __onVentilatorCumulative__<br>
_Field type:integer_<br>
Cumulative on ventilator/Ever on ventilator<br>
Total number of individuals who have ever been hospitalized under advanced ventilation with COVID-19. Definitions vary by state / territory, and it is not always clear whether pediatric patients are included in this metric. Where possible, we report patients on ventilation with confirmed or suspected COVID-19 cases.<br>
_Returns : null, if no data is available_<br>

29. __onVentilatorCurrently__<br>
_Field type:integer_<br>
Currently on ventilator/Now on ventilator<br>
Individuals who are currently hospitalized under advanced ventilation with COVID-19. Definitions vary by state / territory, and it is not always clear whether pediatric patients are included in this metric. Where possible, we report patients on ventilation with confirmed or suspected COVID-19 cases.<br>
_Returns : null, if no data is available_<br>

30. __pending__<br>
_Field type:integer_<br>
Pending<br>
Total number of viral tests that have not been completed as reported by the state or territory.<br>
_Returns : null, if no data is available_<br>

31. __posNeg__<br>
_Field type:integer_<br>
Deprecated<br>
_Returns : null, if no data is available_<br>

32. __positive__<br>
_Field type:integer_<br>
Cases (confirmed plus probable)<br>
Total number of confirmed plus probable cases of COVID-19 reported by the state or territory, ideally per the August 5, 2020 CSTE case definition. Some states are following the older April 5th, 2020 CSTE case definition or using their own custom definitions. Not all states and territories report probable cases. If a state is not reporting probable cases, this field will just represent confirmed cases.<br>
_Returns : null, if no data is available_<br>

33. __positiveCasesViral__<br>
_Field type:integer_<br>
Confirmed Cases<br>
Total number of unique people with a positive PCR or other approved nucleic acid amplification test (NAAT), as reported by the state or territory. This is equivalent to a confirmed case as per the CSTE case definitions of August 5th, 2020 and April 5th, 2020. If we discover a jurisdiction is labeling cases as confirmed using other evidence (e.g. positive antigen tests), we will annotate that state or territory’s data accordingly.<br>
_Returns : null, if no data is available_<br>

34. __positiveIncrease__<br>
_Field type:integer_<br>
New cases<br>
The daily increase in API field positive, which measures Cases (confirmed plus probable) calculated based on the previous day’s value.<br>
_Returns : null, if no data is available_<br>

35. __positiveScore__<br>
_Field type:integer_<br>
Deprecated<br>
_Returns : null, if no data is available_<br>

36. __positiveTestsAntibody__<br>
_Field type:integer_<br>
Positive antibody tests (specimens)<br>
Total number of completed antibody tests that return positive as reported by the state or territory.<br>
_Returns : null, if no data is available_

37. __positiveTestsAntigen__<br>
_Field type:integer_<br>
Positive antigen tests (specimens)<br>
Total number of completed antigen tests that return positive as reported by the state or territory.<br>
_Returns : null, if no data is available_

38. __positiveTestsPeopleAntibody__<br>
_Field type:integer_<br>
Positive antibody tests (people)<br>
The total number of unique people with completed antibody tests that return positive as reported by the state or territory.<br>
_Returns : null, if no data is available_<br>

39. __positiveTestsPeopleAntigen__<br>
_Field type:integer_<br>
Positive antigen tests (people)<br>
Total number of unique people with a completed antigen test that returned positive as reported by the state or territory.<br>
_Returns : null, if no data is available_<br>

40. __positiveTestsViral__<br>
_Field type:integer_<br>
Positive PCR tests (specimens)<br>
Total number of completed PCR tests (or specimens tested) that return positive as reported by the state or territory. If we discover that a jurisdiction is including antigen tests in this metric, we will annotate that state or territory’s data accordingly.<br>
_Returns : null, if no data is available_<br>

41. __probableCases__<br>
_Field type:integer_<br>
Probable Cases<br>
Total number of probable cases of COVID-19 as reported by the state or territory, ideally per the August 5, 2020 CSTE case definition. By this definition, a probable case is someone who tests positive via antigen without a positive PCR or other approved nucleic acid amplification test (NAAT), someone with clinical and epidemiological evidence of COVID-19 infection with no confirmatory laboratory testing performed for SARS-CoV-2, or someone with COVID-19 listed on their death certificate with no confirmatory laboratory testing performed for SARS-CoV-2. Some states are following the older April 5th, 2020 CSTE case definition or using their own custom definitions.<br>
_Returns : null, if no data is available_<br>

42. __recovered__<br>
_Field type:integer_<br>
Recovered<br>
Total number of people that are identified as recovered from COVID-19. States provide very disparate definitions on what constitutes a “recovered” COVID-19 case. Types of “recovered” cases include those who are discharged from hospitals, released from isolation after meeting CDC guidance on symptoms cessation, or those who have not been identified as fatalities after a number of days (30 or more) post disease onset. Specifics vary for each state or territory.<br>
_Returns : null, if no data is available_<br>

43. __score__<br>
_Field type:integer_<br>
Deprecated<br>
_Returns : null, if no data is available_<br>

44. __state__<br>
_Field type:string_<br>
State (or territory)<br>
Two-letter abbreviation for the state or territory.<br>

45. __total__<br>
_Field type:integer_<br>
Deprecated<br>
_Returns :  null, if no data is available_

46. __totalTestEncountersViral__<br>
_Field type:integer_<br>
Total PCR tests (test encounters)<br>
Total number of people tested per day via PCR testing as reported by the state or territory. The count for this metric is incremented up by one for each day on which an individual person is tested, no matter how many specimens are collected from that person on that day. If an individual person is tested twice a day on three different days, this count will increment up by three. If we discover that a jurisdiction is including antigen tests in this metric, we will annotate that state or territory’s data accordingly.<br>
_Returns : null, if no data is available_<br>

47. __totalTestResults__<br>
_Field type:integer_<br>
Total test results<br>
At the national level, this metric is a summary statistic which, because of the variation in test reporting methods, is at best an estimate of US viral (PCR) testing. Some states/territories report tests in units of test encounters, some report tests in units of specimens, and some report tests in units of unique people. Moreover, some jurisdictions include antigen tests in their total test counts without reporting a separate total of viral (PCR) tests. Therefore, this value is an aggregate calculation of heterogeneous figures. Please consult each state or territory’s individual data page to see whether that jurisdiction lumps antigen and PCR tests together and to see what units that jurisdiction uses for test reporting.<br>
In most states, the totalTestResults field is currently computed by adding positive and negative values because, historically, some states do not report totals, and to work around different reporting cadences for cases and tests. In Colorado, Delaware, the District of Columbia, Florida, Hawaii, Minnesota, Nevada, New York, North Dakota, Pennsylvania, Rhode Island, Virginia, Washington, and Wisconsin, where reliable testing encounters figures are available with a complete time series, we directly report those figures in this field. In Alaska, America Samoa, Arizona, Arkansas, California, Georgia, Indiana, Kentucky, Maryland, Massachusetts, Missouri, Nebraska, New Hampshire, Ohio, Oregon, Texas, Utah, Vermont, and Wyoming, where reliable specimens figures are available with a complete time series, we directly report those figures in this field. In Alabama, Idaho, and South Dakota, where reliable unique people figures are available with a complete time series, we directly report those figures in this field. We are in the process of switching all states over to use directly reported total figures, using a policy of preferring testing encounters, specimens, and people, in that order.<br>
_Returns : null, if no data is available_<br>


48. __totalTestResultsIncrease__<br>
_Field type:integer_<br>
New tests<br>
Daily increase in totalTestResults, calculated from the previous day’s value. This calculation includes all the caveats associated with Total tests/totalTestResults, and we recommend against using it at the state/territory level.<br>
_Returns : null, if no data is available_<br>

49. __totalTestResultsSource__<br>
_Field type:string_<br>
Indicates which field is being used for total test results. If it is posNeg, then it is calculated by adding all positive and negative values.<br>

50. __totalTestsAntibody__<br>
_Field type:integer_<br>
Total antibody tests (specimens)<br>
Total number of completed antibody tests as reported by the state or territory.<br>
_Returns : null, if no data is available_<br>

51. __totalTestsAntigen__<br>
_Field type:integer_<br>
Total antigen tests (specimens)<br>
Total number of completed antigen tests, as reported by the state or territory.<br>
_Returns : null, if no data is available_<br>

52. __totalTestsPeopleAntibody__<br>
_Field type:integer_<br>
Total antibody tests (people)<br>
The total number of unique people who have been tested at least once via antibody testing as reported by the state or territory.<br>
_Returns : null, if no data is available_<br>

53. __totalTestsPeopleAntigen__<br>
_Field type:integer_<br>
Total antigen tests (people)<br>
Total number of unique people who have been tested at least once via antigen testing, as reported by the state or territory.<br>
_Returns : null, if no data is available_

54. __totalTestsPeopleViral__<br>
_Field type:integer_<br>
Total PCR tests (people)<br>
Total number of unique people tested at least once via PCR testing, as reported by the state or territory. The count for this metric is incremented up only the first time an individual person is tested and their result is reported. Future tests of the same person will not be added to this count.<br>
In the case where the state only provides negative cases, this field is calculated as the summation of people who tested positive (“Positive Cases (People”) and the number of people who tested negative (“Negative (People or Cases)”). If we discover that a jurisdiction is including antigen tests in this metric, we will annotate that state or territory’s data accordingly.<br>
_Returns : null, if no data is available_<br>


55. __totalTestsViral__<br>
_Field type:integer_<br>
Total PCR tests (specimens)<br>
Total number of PCR tests (or specimens tested) as reported by the state or territory. The count for this metric is incremented up each time a specimen is tested and the result is reported. If we discover that a jurisdiction is including antigen tests in this metric, we will annotate that state or territory’s data accordingly. For states with ambiguous annotations, we have assigned their total tests to this category; these states and territories are identified in the new API field covidTrackingProjectPreferredTotalTestUnits as having “Unclear units.”<br>
_Returns : null, if no data is available_

In [6]:
df.corr() # Checking the correlation values between the different features

date  positive  probableCases  negative  \
date                         1.000000  0.405193       0.479608  0.358384   
positive                     0.405193  1.000000       0.901992  0.861105   
probableCases                0.479608  0.901992       1.000000  0.618816   
negative                     0.358384  0.861105       0.618816  1.000000   
pending                      0.051330  0.228040       0.709812  0.254607   
totalTestResults             0.344970  0.918980       0.760841  0.915811   
hospitalizedCurrently        0.181355  0.787494       0.797650  0.611212   
hospitalizedCumulative       0.238533  0.810706       0.715020  0.856262   
inIcuCurrently               0.108407  0.737543       0.786963  0.465121   
inIcuCumulative              0.379155  0.870195       0.752833  0.854938   
onVentilatorCurrently        0.128849  0.544824       0.456695  0.322258   
onVentilatorCumulative       0.467084  0.809313       0.893624  0.913416   
recovered                    0.365103  0.868672       0.867068  0.691156   
death                        0.272942  0.846884       0.708197  0.824405   
hospitalized                 0.238533  0.810706       0.715020  0.856262   
totalTestsViral              0.334987  0.938540       0.765646  0.924850   
positiveTestsViral           0.384954  0.985490       0.905259  0.866824   
negativeTestsViral           0.363208  0.959147       0.828080  0.951887   
positiveCasesViral           0.370405  0.996986       0.873533  0.848241   
deathConfirmed               0.323711  0.774637       0.437406  0.786581   
deathProbable                0.387327  0.590889       0.508988  0.541846   
totalTestEncountersViral     0.378765  0.907070       0.861294  0.983612   
totalTestsPeopleViral        0.319400  0.954244       0.771645  0.998754   
totalTestsAntibody           0.231403  0.903174       0.735247  0.865249   
positiveTestsAntibody        0.332863  0.873546       0.771055  0.743027   
negativeTestsAntibody        0.164206  0.913569       0.702709  0.919803   
totalTestsPeopleAntibody     0.157295  0.911048       0.627095  0.938015   
positiveTestsPeopleAntibody  0.291788  0.951632       0.726455  0.941740   
negativeTestsPeopleAntibody  0.232095  0.932915       0.768986  0.971995   
totalTestsPeopleAntigen      0.716104  0.843731       0.580138  0.820048   
positiveTestsPeopleAntigen   0.823402  0.917104       0.999777  0.780839   
totalTestsAntigen            0.536398  0.791766       0.925872  0.394668   
positiveTestsAntigen         0.601337  0.806491       0.900742  0.603542   
fips                         0.001362 -0.140291      -0.078626 -0.197155   
positiveIncrease             0.232747  0.797903       0.696148  0.679624   
negativeIncrease             0.060034  0.100782       0.039651  0.290419   
total                        0.347057  0.564459       0.455598  0.998699   
totalTestResultsIncrease     0.196105  0.783415       0.504699  0.734474   
posNeg                       0.347089  0.564435       0.455513  0.998708   
deathIncrease                0.205265  0.633014       0.634529  0.449624   
hospitalizedIncrease         0.039087  0.094815       0.105033  0.105550   
commercialScore                   NaN       NaN            NaN       NaN   
negativeRegularScore              NaN       NaN            NaN       NaN   
negativeScore                     NaN       NaN            NaN       NaN   
positiveScore                     NaN       NaN            NaN       NaN   
score                             NaN       NaN            NaN       NaN   
grade                             NaN       NaN            NaN       NaN   

                              pending  totalTestResults  \
date                         0.051330          0.344970   
positive                     0.228040          0.918980   
probableCases                0.709812          0.760841   
negative                     0.254607          0.915811   
pending                      1.000000          0.223423   
totalTestResults             0

In [7]:
# Checking the correlation between the death vs all the features only
correlation_df = pd.DataFrame(df.corr()['death']).reset_index()
correlation_df

index     death
0                          date  0.272942
1                      positive  0.846884
2                 probableCases  0.708197
3                      negative  0.824405
4                       pending  0.208730
5              totalTestResults  0.853438
6         hospitalizedCurrently  0.658345
7        hospitalizedCumulative  0.941685
8                inIcuCurrently  0.600416
9               inIcuCumulative  0.956904
10        onVentilatorCurrently  0.458481
11       onVentilatorCumulative  0.884807
12                    recovered  0.562122
13                        death  1.000000
14                 hospitalized  0.941685
15              totalTestsViral  0.867394
16           positiveTestsViral  0.935769
17           negativeTestsViral  0.952699
18           positiveCasesViral  0.874143
19               deathConfirmed  0.997818
20                deathProbable  0.803197
21     totalTestEncountersViral  0.855564
22        totalTestsPeopleViral  0.920230
23           totalTestsAntibody  0.947593
24        positiveTestsAntibody  0.874169
25        negativeTestsAntibody  0.942658
26     totalTestsPeopleAntibody  0.873022
27  positiveTestsPeopleAntibody  0.924089
28  negativeTestsPeopleAntibody  0.960674
29      totalTestsPeopleAntigen  0.709044
30   positiveTestsPeopleAntigen  0.868064
31            totalTestsAntigen  0.711943
32         positiveTestsAntigen  0.713970
33                         fips -0.148165
34             positiveIncrease  0.625875
35             negativeIncrease  0.119605
36                        total  0.556296
37     totalTestResultsIncrease  0.771394
38                       posNeg  0.556301
39                deathIncrease  0.546197
40         hospitalizedIncrease  0.103311
41              commercialScore       NaN
42         negativeRegularScore       NaN
43                negativeScore       NaN
44                positiveScore       NaN
45                        score       NaN
46                        grade       NaN

In [8]:
# Selecting the features based on the correlation value.
feature_columns = []
for i,j in zip(correlation_df['index'],correlation_df['death']):
    if abs(j) > 0.6 and abs(j) < 0.99:
        feature_columns.append(i)

In [9]:
feature_columns

['positive',
 'probableCases',
 'negative',
 'totalTestResults',
 'hospitalizedCurrently',
 'hospitalizedCumulative',
 'inIcuCurrently',
 'inIcuCumulative',
 'onVentilatorCumulative',
 'hospitalized',
 'totalTestsViral',
 'positiveTestsViral',
 'negativeTestsViral',
 'positiveCasesViral',
 'deathProbable',
 'totalTestEncountersViral',
 'totalTestsPeopleViral',
 'totalTestsAntibody',
 'positiveTestsAntibody',
 'negativeTestsAntibody',
 'totalTestsPeopleAntibody',
 'positiveTestsPeopleAntibody',
 'negativeTestsPeopleAntibody',
 'totalTestsPeopleAntigen',
 'positiveTestsPeopleAntigen',
 'totalTestsAntigen',
 'positiveTestsAntigen',
 'positiveIncrease',
 'totalTestResultsIncrease']

In [10]:
feature_columns.append('death')

In [11]:
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 19261 entries, 0 to 19260
Data columns (total 55 columns):
 #   Column                       Non-Null Count  Dtype  
---  ------                       --------------  -----  
 0   date                         19261 non-null  int64  
 1   state                        19261 non-null  object 
 2   positive                     19072 non-null  float64
 3   probableCases                8246 non-null   float64
 4   negative                     15334 non-null  float64
 5   pending                      2004 non-null   float64
 6   totalTestResultsSource       19261 non-null  object 
 7   totalTestResults             19159 non-null  float64
 8   hospitalizedCurrently        15907 non-null  float64
 9   hospitalizedCumulative       11865 non-null  float64
 10  inIcuCurrently               10400 non-null  float64
 11  inIcuCumulative              3468 non-null   float64
 12  onVentilatorCurrently        8258 non-null   float64
 13  onVentilatorCumu

In [12]:
# Seems the total tests done = Positive + Negative
# However there are NAN values in negative and positive columns.
# positive = 19016
# negative = 15289
# total = 19205
# The data is lagging some positive cases, many negative cases
print(df['positive'].isna().sum())
print(df['negative'].isna().sum())
print(df['total'].isna().sum())

189
3927
0


In [13]:
# Some analysis on NAN values in negative columns
df[df['negative'].isna()][['positive', 'negative', 'total']]
# It shows that, if the negative value is nan then, the total equals positive

positive  negative    total
0        53694.0       NaN    53694
5      3346340.0       NaN  3346340
8        38136.0       NaN    38136
11      943695.0       NaN   943695
13       27195.0       NaN    27195
...          ...       ...      ...
19256        0.0       NaN        0
19257        0.0       NaN        0
19258        0.0       NaN        0
19259        0.0       NaN        0
19260        NaN       NaN        0

[3927 rows x 3 columns]

In [14]:
# Some analysis on NAN values in positive columns
df[df['positive'].isna()][['positive', 'negative', 'total']]

# It shows that, if the positive value is nan then, the total is zero and negative is also nan

positive  negative  total
17643       NaN       NaN      0
17699       NaN       NaN      0
17755       NaN       NaN      0
17811       NaN       NaN      0
17867       NaN       NaN      0
...         ...       ...    ...
19244       NaN       NaN      0
19246       NaN       NaN      0
19248       NaN       NaN      0
19250       NaN       NaN      0
19260       NaN       NaN      0

[189 rows x 3 columns]

In [15]:
# Dropping all the columns with total = 0, which means the data is not available
df = df[df['total']!=0]

In [16]:
# Based on above filtering, checking out Nan values in positive columns
df[df['positive'].isna()][['positive', 'negative', 'total']]

positive  negative  total
18633       NaN       NaN     14
18684       NaN       NaN     14
18735       NaN       NaN      9
18786       NaN       NaN      9
18837       NaN       NaN      6
18888       NaN       NaN      2
18939       NaN       NaN      1

In [17]:
# Based on above filtering, checking out Nan values in negative columns
df[df['negative'].isna()][['positive', 'negative', 'total']]

positive  negative    total
0        53694.0       NaN    53694
5      3346340.0       NaN  3346340
8        38136.0       NaN    38136
11      943695.0       NaN   943695
13       27195.0       NaN    27195
...          ...       ...      ...
19249        2.0       NaN        2
19251        2.0       NaN        2
19252        2.0       NaN        2
19253        1.0       NaN        1
19254        1.0       NaN        1

[3680 rows x 3 columns]

In [18]:
# Looking at the 3669 scenarios, all the positives = total when negative is nan, 
#thus replacing all the nan of negative with 0
df['negative'].fillna(0, inplace=True)

In [19]:
df[df['negative'].isna()][['positive', 'negative', 'total']]

Empty DataFrame
Columns: [positive, negative, total]
Index: []

In [20]:
# Placing all the values of total to positive, when the positive is nan
df['positive'] = df.apply(lambda row: row['total'] if np.isnan(row['positive']) else row['positive'], axis=1)

In [21]:
df[df['positive'].isna()][['positive', 'negative', 'total']]

Empty DataFrame
Columns: [positive, negative, total]
Index: []

In [22]:
df.info()

<class 'pandas.core.frame.DataFrame'>
Int64Index: 19009 entries, 0 to 19254
Data columns (total 55 columns):
 #   Column                       Non-Null Count  Dtype  
---  ------                       --------------  -----  
 0   date                         19009 non-null  int64  
 1   state                        19009 non-null  object 
 2   positive                     19009 non-null  float64
 3   probableCases                8246 non-null   float64
 4   negative                     19009 non-null  float64
 5   pending                      2002 non-null   float64
 6   totalTestResultsSource       19009 non-null  object 
 7   totalTestResults             18948 non-null  float64
 8   hospitalizedCurrently        15907 non-null  float64
 9   hospitalizedCumulative       11861 non-null  float64
 10  inIcuCurrently               10400 non-null  float64
 11  inIcuCumulative              3468 non-null   float64
 12  onVentilatorCurrently        8258 non-null   float64
 13  onVentilatorCumu

### Lets confirm, that the model, we want should be able to predict the death cases given the total tests done, and several other relevant features
### The relevant features that should be used for prediction are as follows which are based on correlation values:

_Most of the features are dropped based on the number of non-null counts, since, 
the total number of the values in such features are very less, thus appending zeros or other values in such features
will deviate the performance and the actual data relevancy, thus will try predicting on the basis of highly available 
feature columns and the relevant feature columns only._

------
<u>Target</u> is calculating the total **deaths** based on the above features.

In [23]:
# Selecting only the above mentioned features for further processing
new_df = df[feature_columns]
new_df.head()

positive  probableCases   negative  totalTestResults  \
0   53694.0            NaN        0.0         1555532.0   
1  473348.0       101577.0  1820080.0         2191851.0   
2  307373.0        62960.0  2290742.0         2535155.0   
3       0.0            NaN     2140.0            2140.0   
4  782887.0        51100.0  2820314.0         6993662.0   

   hospitalizedCurrently  hospitalizedCumulative  inIcuCurrently  \
0                   39.0                  1219.0             NaN   
1                 1524.0                 43383.0             NaN   
2                  777.0                 14099.0           274.0   
3                    NaN                     NaN             NaN   
4                 2853.0                 54713.0           828.0   

   inIcuCumulative  onVentilatorCumulative  hospitalized  totalTestsViral  \
0              NaN                     NaN        1219.0        1555532.0   
1           2577.0                  1460.0       43383.0              NaN   
2              NaN                  1458.0       14099.0        2535155.0   
3              NaN                     NaN           NaN           2140.0   
4              NaN                     NaN       54713.0        6993662.0   

   positiveTestsViral  negativeTestsViral  positiveCasesViral  deathProbable  \
0             64862.0           1488905.0                 NaN            NaN   
1                 NaN                 NaN            371771.0         1770.0   
2                 NaN           2290742.0            244413.0         1025.0   
3                 NaN                 NaN                 0.0            NaN   
4                 NaN                 NaN            731787.0         1568.0   

   totalTestEncountersViral  totalTestsPeopleViral  totalTestsAntibody  \
0                       NaN                    NaN                 NaN   
1                       NaN              2191851.0                 NaN   
2                       NaN                    NaN                 NaN   
3                       NaN                    NaN                 NaN   
4                       NaN              3552101.0            423799.0   

   positiveTestsAntibody  negativeTestsAntibody  totalTestsPeopleAntibody  \
0                    NaN                    NaN                       NaN   
1                    NaN                    NaN                  109716.0   
2                    NaN                    NaN                       NaN   
3                    NaN                    NaN                       NaN   
4                    NaN                    NaN                       NaN   

   positiveTestsPeopleAntibody  negativeTestsPeopleAntibody  \
0                          NaN                          NaN   
1                          NaN                          NaN   
2                          NaN                          NaN   
3                          NaN                          NaN   
4                          NaN                          NaN   

   totalTestsPeopleAntigen  positiveTestsPeopleAntigen  totalTestsAntigen  \
0                      NaN                         NaN                NaN   
1                      NaN                         NaN                NaN   
2                 391322.0                     73931.0                NaN   
3                      NaN                         NaN                NaN   
4                      NaN                         NaN                NaN   

   positiveTestsAntigen  positiveIncrease  totalTestResultsIncrease    death  
0                   NaN               415                     18621    279.0  
1                   NaN               925                      4522   8523.0  
2                   NaN               637                      5830   5106.0  
3                   NaN                 0                         0      0.0  
4                   NaN              2250                     11514  14055.0

In [24]:
nan_filter_df = pd.DataFrame(new_df.isna().sum()).reset_index()
nan_filter_df

index      0
0                      positive      0
1                 probableCases  10763
2                      negative      0
3              totalTestResults     61
4         hospitalizedCurrently   3102
5        hospitalizedCumulative   7148
6                inIcuCurrently   8609
7               inIcuCumulative  15541
8        onVentilatorCumulative  17824
9                  hospitalized   7148
10              totalTestsViral   6382
11           positiveTestsViral  11712
12           negativeTestsViral  14610
13           positiveCasesViral   5540
14                deathProbable  12189
15     totalTestEncountersViral  14205
16        totalTestsPeopleViral  10514
17           totalTestsAntibody  14660
18        positiveTestsAntibody  15667
19        negativeTestsAntibody  17630
20     totalTestsPeopleAntibody  17303
21  positiveTestsPeopleAntibody  18023
22  negativeTestsPeopleAntibody  18118
23      totalTestsPeopleAntigen  18145
24   positiveTestsPeopleAntigen  18457
25            totalTestsAntigen  16198
26         positiveTestsAntigen  17173
27             positiveIncrease      0
28     totalTestResultsIncrease      0
29                        death    670

In [25]:
nan_filter_df.columns

Index(['index', 0], dtype='object')

In [26]:
# Selecting feature column based on NAN values less than 1000
nan_filter_columns = []
for i,j in zip(nan_filter_df['index'], nan_filter_df[0]):
    if j < 10000:
        nan_filter_columns.append(i)

In [27]:
nan_filter_columns

['positive',
 'negative',
 'totalTestResults',
 'hospitalizedCurrently',
 'hospitalizedCumulative',
 'inIcuCurrently',
 'hospitalized',
 'totalTestsViral',
 'positiveCasesViral',
 'positiveIncrease',
 'totalTestResultsIncrease',
 'death']

In [29]:
new_df[nan_filter_columns].corr()['death']

positive                    0.846876
negative                    0.454630
totalTestResults            0.853324
hospitalizedCurrently       0.658345
hospitalizedCumulative      0.941678
inIcuCurrently              0.600416
hospitalized                0.941678
totalTestsViral             0.867336
positiveCasesViral          0.874104
positiveIncrease            0.625605
totalTestResultsIncrease    0.771199
death                       1.000000
Name: death, dtype: float64

In [30]:
new_df.head()

positive  probableCases   negative  totalTestResults  \
0   53694.0            NaN        0.0         1555532.0   
1  473348.0       101577.0  1820080.0         2191851.0   
2  307373.0        62960.0  2290742.0         2535155.0   
3       0.0            NaN     2140.0            2140.0   
4  782887.0        51100.0  2820314.0         6993662.0   

   hospitalizedCurrently  hospitalizedCumulative  inIcuCurrently  \
0                   39.0                  1219.0             NaN   
1                 1524.0                 43383.0             NaN   
2                  777.0                 14099.0           274.0   
3                    NaN                     NaN             NaN   
4                 2853.0                 54713.0           828.0   

   inIcuCumulative  onVentilatorCumulative  hospitalized  totalTestsViral  \
0              NaN                     NaN        1219.0        1555532.0   
1           2577.0                  1460.0       43383.0              NaN   
2              NaN                  1458.0       14099.0        2535155.0   
3              NaN                     NaN           NaN           2140.0   
4              NaN                     NaN       54713.0        6993662.0   

   positiveTestsViral  negativeTestsViral  positiveCasesViral  deathProbable  \
0             64862.0           1488905.0                 NaN            NaN   
1                 NaN                 NaN            371771.0         1770.0   
2                 NaN           2290742.0            244413.0         1025.0   
3                 NaN                 NaN                 0.0            NaN   
4                 NaN                 NaN            731787.0         1568.0   

   totalTestEncountersViral  totalTestsPeopleViral  totalTestsAntibody  \
0                       NaN                    NaN                 NaN   
1                       NaN              2191851.0                 NaN   
2                       NaN                    NaN                 NaN   
3                       NaN                    NaN                 NaN   
4                       NaN              3552101.0            423799.0   

   positiveTestsAntibody  negativeTestsAntibody  totalTestsPeopleAntibody  \
0                    NaN                    NaN                       NaN   
1                    NaN                    NaN                  109716.0   
2                    NaN                    NaN                       NaN   
3                    NaN                    NaN                       NaN   
4                    NaN                    NaN                       NaN   

   positiveTestsPeopleAntibody  negativeTestsPeopleAntibody  \
0                          NaN                          NaN   
1                          NaN                          NaN   
2                          NaN                          NaN   
3                          NaN                          NaN   
4                          NaN                          NaN   

   totalTestsPeopleAntigen  positiveTestsPeopleAntigen  totalTestsAntigen  \
0                      NaN                         NaN                NaN   
1                      NaN                         NaN                NaN   
2                 391322.0                     73931.0                NaN   
3                      NaN                         NaN                NaN   
4                      NaN                         NaN                NaN   

   positiveTestsAntigen  positiveIncrease  totalTestResultsIncrease    death  
0                   NaN               415                     18621    279.0  
1                   NaN               925                      4522   8523.0  
2                   NaN               637                      5830   5106.0  
3                   NaN                 0                         0      0.0  
4                   NaN              2250                     11514  14055.0

In [32]:
# new_df['date'] =  pd.to_datetime(new_df['date'], format='%Y%m%d')

In [33]:
new_df['death'].isna().sum()

670

In [36]:
# Lets drop all the rows with death = nan
new_df[new_df['death'].notna()]['death'].isna().sum()

0

In [37]:
new_df = new_df[new_df['death'].notna()]
new_df.head()

positive  probableCases   negative  totalTestResults  \
0   53694.0            NaN        0.0         1555532.0   
1  473348.0       101577.0  1820080.0         2191851.0   
2  307373.0        62960.0  2290742.0         2535155.0   
3       0.0            NaN     2140.0            2140.0   
4  782887.0        51100.0  2820314.0         6993662.0   

   hospitalizedCurrently  hospitalizedCumulative  inIcuCurrently  \
0                   39.0                  1219.0             NaN   
1                 1524.0                 43383.0             NaN   
2                  777.0                 14099.0           274.0   
3                    NaN                     NaN             NaN   
4                 2853.0                 54713.0           828.0   

   inIcuCumulative  onVentilatorCumulative  hospitalized  totalTestsViral  \
0              NaN                     NaN        1219.0        1555532.0   
1           2577.0                  1460.0       43383.0              NaN   
2              NaN                  1458.0       14099.0        2535155.0   
3              NaN                     NaN           NaN           2140.0   
4              NaN                     NaN       54713.0        6993662.0   

   positiveTestsViral  negativeTestsViral  positiveCasesViral  deathProbable  \
0             64862.0           1488905.0                 NaN            NaN   
1                 NaN                 NaN            371771.0         1770.0   
2                 NaN           2290742.0            244413.0         1025.0   
3                 NaN                 NaN                 0.0            NaN   
4                 NaN                 NaN            731787.0         1568.0   

   totalTestEncountersViral  totalTestsPeopleViral  totalTestsAntibody  \
0                       NaN                    NaN                 NaN   
1                       NaN              2191851.0                 NaN   
2                       NaN                    NaN                 NaN   
3                       NaN                    NaN                 NaN   
4                       NaN              3552101.0            423799.0   

   positiveTestsAntibody  negativeTestsAntibody  totalTestsPeopleAntibody  \
0                    NaN                    NaN                       NaN   
1                    NaN                    NaN                  109716.0   
2                    NaN                    NaN                       NaN   
3                    NaN                    NaN                       NaN   
4                    NaN                    NaN                       NaN   

   positiveTestsPeopleAntibody  negativeTestsPeopleAntibody  \
0                          NaN                          NaN   
1                          NaN                          NaN   
2                          NaN                          NaN   
3                          NaN                          NaN   
4                          NaN                          NaN   

   totalTestsPeopleAntigen  positiveTestsPeopleAntigen  totalTestsAntigen  \
0                      NaN                         NaN                NaN   
1                      NaN                         NaN                NaN   
2                 391322.0                     73931.0                NaN   
3                      NaN                         NaN                NaN   
4                      NaN                         NaN                NaN   

   positiveTestsAntigen  positiveIncrease  totalTestResultsIncrease    death  
0                   NaN               415                     18621    279.0  
1                   NaN               925                      4522   8523.0  
2                   NaN               637                      5830   5106.0  
3                   NaN                 0                         0      0.0  
4                   NaN              2250                     11514  14055.0

In [38]:
# Lets refill all the Nans with zeros
new_df.fillna(0, inplace=True)

/Users/adarsh/Desktop/Machine Learning/Projects/Assignment 1 Covid Dataset/.venv/lib/python3.8/site-packages/pandas/core/frame.py:4459: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  return super().fillna(


In [39]:
new_df.head()

positive  probableCases   negative  totalTestResults  \
0   53694.0            0.0        0.0         1555532.0   
1  473348.0       101577.0  1820080.0         2191851.0   
2  307373.0        62960.0  2290742.0         2535155.0   
3       0.0            0.0     2140.0            2140.0   
4  782887.0        51100.0  2820314.0         6993662.0   

   hospitalizedCurrently  hospitalizedCumulative  inIcuCurrently  \
0                   39.0                  1219.0             0.0   
1                 1524.0                 43383.0             0.0   
2                  777.0                 14099.0           274.0   
3                    0.0                     0.0             0.0   
4                 2853.0                 54713.0           828.0   

   inIcuCumulative  onVentilatorCumulative  hospitalized  totalTestsViral  \
0              0.0                     0.0        1219.0        1555532.0   
1           2577.0                  1460.0       43383.0              0.0   
2              0.0                  1458.0       14099.0        2535155.0   
3              0.0                     0.0           0.0           2140.0   
4              0.0                     0.0       54713.0        6993662.0   

   positiveTestsViral  negativeTestsViral  positiveCasesViral  deathProbable  \
0             64862.0           1488905.0                 0.0            0.0   
1                 0.0                 0.0            371771.0         1770.0   
2                 0.0           2290742.0            244413.0         1025.0   
3                 0.0                 0.0                 0.0            0.0   
4                 0.0                 0.0            731787.0         1568.0   

   totalTestEncountersViral  totalTestsPeopleViral  totalTestsAntibody  \
0                       0.0                    0.0                 0.0   
1                       0.0              2191851.0                 0.0   
2                       0.0                    0.0                 0.0   
3                       0.0                    0.0                 0.0   
4                       0.0              3552101.0            423799.0   

   positiveTestsAntibody  negativeTestsAntibody  totalTestsPeopleAntibody  \
0                    0.0                    0.0                       0.0   
1                    0.0                    0.0                  109716.0   
2                    0.0                    0.0                       0.0   
3                    0.0                    0.0                       0.0   
4                    0.0                    0.0                       0.0   

   positiveTestsPeopleAntibody  negativeTestsPeopleAntibody  \
0                          0.0                          0.0   
1                          0.0                          0.0   
2                          0.0                          0.0   
3                          0.0                          0.0   
4                          0.0                          0.0   

   totalTestsPeopleAntigen  positiveTestsPeopleAntigen  totalTestsAntigen  \
0                      0.0                         0.0                0.0   
1                      0.0                         0.0                0.0   
2                 391322.0                     73931.0                0.0   
3                      0.0                         0.0                0.0   
4                      0.0                         0.0                0.0   

   positiveTestsAntigen  positiveIncrease  totalTestResultsIncrease    death  
0                   0.0               415                     18621    279.0  
1                   0.0               925                      4522   8523.0  
2                   0.0               637                      5830   5106.0  
3                   0.0                 0                         0      0.0  
4                   0.0              2250                     11514  14055.0

### Dividing data into train, and test sets

In [40]:
# Since the dataset is ordered in chronological manner, 
# lets put 70% of past data as training data and latest 30% as test data
mask = round(0.3*len(new_df))
print(mask)
test = new_df[:mask]
train = new_df[mask:]

5502


In [41]:
len(train)

12837

In [42]:
len(test)

5502

In [43]:
# resetting the index 
train.reset_index(inplace=True, drop=True)

In [44]:
train_features = list(set(train.columns) - set(['death']))
# print(train_features)
X_train = train[train_features]
y_train = train["death"]
X_test = test[train_features]
y_test = test["death"]

### Simple Linear regression without any feature manipulations

In [45]:
from sklearn.linear_model import LinearRegression
from sklearn.metrics import mean_squared_error

In [46]:
simple_regressor = LinearRegression()
simple_regressor.fit(X_train, y_train)

LinearRegression()

In [47]:
predictions = simple_regressor.predict(X_train)

In [48]:
print('RMSE for Simple Linear Regression : ',np.sqrt(mean_squared_error(y_train, predictions)))

RMSE for Simple Linear Regression :  964.6419913821259


#### Standardization of feature columns

In [49]:
from sklearn.preprocessing import StandardScaler

In [50]:
scaler = StandardScaler()

In [51]:
scaler.fit(X_train)

StandardScaler()

In [52]:
scaler.transform(X_train)

array([[17.64087025,  0.45018863, -0.2353503 , ..., -0.30985538,
        -0.40109972, -0.14036875],
       [-0.10548487, -0.02041079, -0.2353503 , ..., -0.30985538,
        -0.4944986 , -0.14036875],
       [-0.10548487,  2.92404917, -0.2353503 , ..., -0.30985538,
         2.81725759, -0.14036875],
       ...,
       [-0.10548487, -0.54582628, -0.2353503 , ..., -0.30985538,
        -0.59225209, -0.14036875],
       [-0.10548487, -0.54584243, -0.2353503 , ..., -0.30985538,
        -0.59225209, -0.14036875],
       [-0.10548487, -0.54585051, -0.2353503 , ..., -0.30985538,
        -0.59225209, -0.14036875]])

In [53]:
standardized_regressor = LinearRegression()
standardized_regressor.fit(scaler.transform(X_train), y_train)

LinearRegression()

In [54]:
standardized_predictions = standardized_regressor.predict(scaler.transform(X_train))

In [55]:
print('RMSE for Standardized Linear Regression : ',np.sqrt(mean_squared_error(y_train, standardized_predictions)))

RMSE for Standardized Linear Regression :  964.6600043779277


Clearly, the above result of RMSE is very bad, saying that the model is quiet underfitting

-------

#### Trying out polynomial features for fitting the data

In [56]:
from sklearn.preprocessing import PolynomialFeatures

#### Polynomial regression with degree 2

In [57]:
poly_2 = PolynomialFeatures(degree = 2) 
X_poly_2 = poly_2.fit_transform(X_train)

poly_scaler_2 = StandardScaler()
poly_scaler_2.fit(X_poly_2)

poly_reg_2 = LinearRegression() 
poly_reg_2.fit(poly_scaler_2.transform(X_poly_2), y_train) 
pred_2 = poly_reg_2.predict(poly_scaler_2.transform(X_poly_2))

print('RMSE for polynomial features with degree 2 Regression :', np.sqrt(mean_squared_error(y_train, pred_2)))

RMSE for polynomial features with degree 2 Regression : 198.08790783336596


Degree 2 for the polynomial feature has reduced the RMSE value by almost 5 times, the model is obviously fitting good

----

#### Polynomial regression with degree 3

In [58]:
poly_3 = PolynomialFeatures(degree = 3) 
X_poly_3 = poly_3.fit_transform(X_train)

poly_scaler_3 = StandardScaler()
poly_scaler_3.fit(X_poly_3)

poly_reg_3 = LinearRegression() 
poly_reg_3.fit(poly_scaler_3.transform(X_poly_3), y_train) 
pred_3 = poly_reg_3.predict(poly_scaler_3.transform(X_poly_3))

print('RMSE for polynomial features with degree 3 Regression :', np.sqrt(mean_squared_error(y_train, pred_3)))

RMSE for polynomial features with degree 3 Regression : 26.560417323798433


### Result analysis

The linear regression with polynomial feature has decreased the RMSE error significantly.
The degree 3 polynomial feature has the lowest RMSE in training data, that is 26.56.

In [59]:
# Lets check if the model is overfitting.
print("RMSE for test data with polynomial regression of degree 3 : ", np.sqrt(mean_squared_error(y_test, 
                                                                                                 poly_reg_3.predict(poly_scaler_3.transform(poly_3.transform(X_test))))))

RMSE for test data with polynomial regression of degree 3 :  1.3675018799820007e+24


It can be clearly seen from the RMSE score that the polynomial regression model of degree 3 is overfitting alot

In [60]:
# Lets check if the model is overfitting in polynomial regression with degree 2.
test_poly_2 = PolynomialFeatures(degree = 2) 
X_test_2 = test_poly_2.fit_transform(X_test)
print("RMSE for test data with polynomial regression of degree 2 : ", np.sqrt(mean_squared_error(y_test, 
                                                                                                 poly_reg_2.predict(poly_scaler_2.transform(poly_2.transform(X_test))))))

RMSE for test data with polynomial regression of degree 2 :  2.0353798479707325e+17


Polynomial regression model of degree 2 is also very overfitting

In [61]:
print("RMSE for test data with standardized regressor : ", np.sqrt(mean_squared_error(y_test, standardized_regressor.predict(scaler.transform(X_test)))))

RMSE for test data with standardized regressor :  5825.747369541201


------

### Applying regularization

#### L2 regularization

#### L2 regularization on linear regression

In [62]:
from sklearn.linear_model import Ridge
ridge_standardized = Ridge(alpha=100.0)

In [63]:
ridge_standardized.fit(scaler.transform(X_train), y_train)

Ridge(alpha=100.0)

In [64]:
print('RMSE on training data for Standardized Ridge Regression : ',np.sqrt(mean_squared_error(y_train, ridge_standardized.predict(scaler.transform(X_train)))))

RMSE on training data for Standardized Ridge Regression :  996.714668427396


In [65]:
print("RMSE for test data with ridge regressor : ", np.sqrt(mean_squared_error(y_test, ridge_standardized.predict(scaler.transform(X_test)))))

RMSE for test data with ridge regressor :  4624.969990451855


As from the result of regularization on standard linear regression, it can be seen that the model performance on test data has improved slightly as compared to the without regularization result.

#### L2 regularization on polynomial regression of degree 2

In [66]:
ridge_poly_2 = Ridge(alpha=100.0)

In [67]:
ridge_poly_2.fit(poly_scaler_2.transform(poly_2.transform(X_train)), y_train)

Ridge(alpha=100.0)

In [68]:
print('RMSE on training data for Standardized Ridge Regression with degree 2 : ',np.sqrt(mean_squared_error(y_train, ridge_poly_2.predict(poly_scaler_2.transform(poly_2.transform(X_train))))))

RMSE on training data for Standardized Ridge Regression with degree 2 :  433.6031367753366


In [69]:
print('RMSE on test data for Standardized Ridge Regression with degree 2 : ',np.sqrt(mean_squared_error(y_test, ridge_poly_2.predict(poly_scaler_2.transform(poly_2.transform(X_test))))))

RMSE on test data for Standardized Ridge Regression with degree 2 :  28954.29828181943


By adding L2 regularization on polynomial regression of order 2, the model performance on test data improved by 1.06\*10^15 to 4272. The addition of L2 regularization on the model has scaled the test error from factor of 10^15 to 1K, which is a huge improvement interms of the model generalizability.

#### L2 regularization on polynomial regression of degree 3

In [70]:
ridge_poly_3 = Ridge(alpha=100.0)

In [71]:
ridge_poly_3.fit(poly_scaler_3.transform(poly_3.transform(X_train)), y_train)

Ridge(alpha=100.0)

In [72]:
print('RMSE on training data for Standardized Ridge Regression with degree 3 : ',np.sqrt(mean_squared_error(y_train, ridge_poly_3.predict(poly_scaler_3.transform(poly_3.transform(X_train))))))

RMSE on training data for Standardized Ridge Regression with degree 3 :  336.5891887210632


In [73]:
print('RMSE on test data for Standardized Ridge Regression with degree 3 : ',np.sqrt(mean_squared_error(y_test, ridge_poly_3.predict(poly_scaler_3.transform(poly_3.transform(X_test))))))

RMSE on test data for Standardized Ridge Regression with degree 3 :  16331811.103061527


By adding L2 regularization on polynomial regression of order 3, the model performance on test data improved by 7.6\*10^16 to 17804. The generalizability of the model has improved very highly. From the result of normal polynomial regression with degree 3 the training error was only 17 however the test error was in 10^16 which is a very high error, and that was showing that the model was overfitting. But by adding regularization the training error increased by a little however the test error reduced from factor of 10^16 to 10^3.
Eventhough, the model performance is not good, but the adding of penalty and its effect on model generalizibility can be clearly seen. 

------

#### L1 regularization

#### L1 regularization on linear regression

In [74]:
from sklearn.linear_model import Lasso
lasso_standardized = Lasso(alpha=100.0)

In [75]:
lasso_standardized.fit(scaler.transform(X_train), y_train)

Lasso(alpha=100.0)

In [76]:
print('RMSE on training data for Standardized lasso Regression : ',np.sqrt(mean_squared_error(y_train, lasso_standardized.predict(scaler.transform(X_train)))))

RMSE on training data for Standardized lasso Regression :  1124.5264059426424


In [77]:
print("RMSE for test data with lasso regressor : ", np.sqrt(mean_squared_error(y_test, lasso_standardized.predict(scaler.transform(X_test)))))

RMSE for test data with lasso regressor :  4452.716485462053


As from the result of regularization on standard linear regression, it can be seen that the model performance on test data has improved slightly as compared to the without regularization result. In addition, the model performance with L1 or L2 regression are quiet comparable

#### L1 regularization on polynomial regression of degree 2

In [78]:
lasso_poly_2 = Lasso(alpha=100.0)

In [79]:
lasso_poly_2.fit(poly_scaler_2.transform(poly_2.transform(X_train)), y_train)

Lasso(alpha=100.0)

In [80]:
print('RMSE on training data for Standardized Lasso Regression with degree 2 : ',np.sqrt(mean_squared_error(y_train, lasso_poly_2.predict(poly_scaler_2.transform(poly_2.transform(X_train))))))

RMSE on training data for Standardized Lasso Regression with degree 2 :  927.5544427946057


In [81]:
print('RMSE on test data for Standardized Lasso Regression with degree 2 : ',np.sqrt(mean_squared_error(y_test, lasso_poly_2.predict(poly_scaler_2.transform(poly_2.transform(X_test))))))

RMSE on test data for Standardized Lasso Regression with degree 2 :  4667.851040008511


By adding L1 regularization on polynomial regression of order 2, the model performance on test data improved by 1.06\*10^15 to 4599. The addition of L1 regularization on the model has scaled the test error from factor of 10^15 to 1K, which is a huge improvement interms of the model generalizability.

#### L1 regularization on polynomial regression of degree 3

In [82]:
lasso_poly_3 = Lasso(alpha=100.0)

In [83]:
lasso_poly_3.fit(poly_scaler_3.transform(poly_3.transform(X_train)), y_train)

Lasso(alpha=100.0)

In [84]:
print('RMSE on training data for Standardized Lasso Regression with degree 3 : ',np.sqrt(mean_squared_error(y_train, lasso_poly_3.predict(poly_scaler_3.transform(poly_3.transform(X_train))))))

RMSE on training data for Standardized Lasso Regression with degree 3 :  884.8006084956604


In [85]:
print('RMSE on test data for Standardized Lasso Regression with degree 3 : ',np.sqrt(mean_squared_error(y_test, lasso_poly_3.predict(poly_scaler_3.transform(poly_3.transform(X_test))))))

RMSE on test data for Standardized Lasso Regression with degree 3 :  13922.537563374512


By adding L1 regularization on polynomial regression of order 3, the model performance on test data improved by 7.6\*10^16 to 5682. The generalizability of the model has improved very highly. From the result of normal polynomial regression with degree 3 the training error was only 17 however the test error was in 10^16 which is a very high error, and that was showing that the model was overfitting. But by adding regularization the training error increased by a little however the test error reduced from factor of 10^16 to 10^3. The L1 regularization has shown better generalizibility than the L2 regularization for the polynomial regression of order 3.
Eventhough, the model performance is not good, but the adding of penalty and its effect on model generalizibility can be clearly seen. 

------

#### L1 and L2 regularization

#### L1/L2 regularization on linear regression

In [86]:
from sklearn.linear_model import ElasticNet

In [87]:
elastic_standardized = ElasticNet(alpha=1.0)

In [88]:
elastic_standardized.fit(scaler.transform(X_train), y_train)

ElasticNet()

In [89]:
print('RMSE on training data for Standardized lasso Regression : ',np.sqrt(mean_squared_error(y_train, elastic_standardized.predict(scaler.transform(X_train)))))

RMSE on training data for Standardized lasso Regression :  1296.8399816414424


In [90]:
print("RMSE for test data with lasso regressor : ", np.sqrt(mean_squared_error(y_test, elastic_standardized.predict(scaler.transform(X_test)))))

RMSE for test data with lasso regressor :  4065.4661008815


As from the result of regularization on standard linear regression, it can be seen that the model performance on test data has improved slightly as compared to the without regularization result. In addition, the model performance with L1, L2, regularization are quiet comparable. However, the L1/L2 regularization with very less penalty has very good performance.

#### L1/L2 regularization on polynomial regression of degree 2

In [91]:
elastic_poly_2 = ElasticNet(alpha=1.0)

In [92]:
elastic_poly_2.fit(poly_scaler_2.transform(poly_2.transform(X_train)), y_train)

/Users/adarsh/Desktop/Machine Learning/Projects/Assignment 1 Covid Dataset/.venv/lib/python3.8/site-packages/sklearn/linear_model/_coordinate_descent.py:530: ConvergenceWarning: Objective did not converge. You might want to increase the number of iterations. Duality gap: 8249554419.003542, tolerance: 22624797.775337458
  model = cd_fast.enet_coordinate_descent(


ElasticNet()

In [93]:
print('RMSE on training data for Standardized Elastic Regression with degree 2 : ',np.sqrt(mean_squared_error(y_train, elastic_poly_2.predict(poly_scaler_2.transform(poly_2.transform(X_train))))))

RMSE on training data for Standardized Elastic Regression with degree 2 :  917.1170279422518


In [94]:
print('RMSE on test data for Standardized Elastic Regression with degree 2 : ',np.sqrt(mean_squared_error(y_test, elastic_poly_2.predict(poly_scaler_2.transform(poly_2.transform(X_test))))))

RMSE on test data for Standardized Elastic Regression with degree 2 :  9261.44547748575


By adding L1/L2 regularization on polynomial regression of order 2, the model performance on test data improved by 1.06\*10^15 to 9373. The addition of L1/L2 regularization on the model has scaled the test error from factor of 10^15 to 10K, which is a huge improvement interms of the model generalizability. However, the model performance is not par as compared to the L1 only and L2 only regularization.

#### L1/L2 regularization on polynomial regression of degree 3

In [95]:
elastic_poly_3 = ElasticNet(alpha=1.0)

In [96]:
elastic_poly_3.fit(poly_scaler_3.transform(poly_3.transform(X_train)), y_train)

/Users/adarsh/Desktop/Machine Learning/Projects/Assignment 1 Covid Dataset/.venv/lib/python3.8/site-packages/sklearn/linear_model/_coordinate_descent.py:530: ConvergenceWarning: Objective did not converge. You might want to increase the number of iterations. Duality gap: 5983917345.734745, tolerance: 22624797.775337458
  model = cd_fast.enet_coordinate_descent(


ElasticNet()

In [97]:
print('RMSE on training data for Standardized Elastic Regression with degree 3 : ',np.sqrt(mean_squared_error(y_train, elastic_poly_3.predict(poly_scaler_3.transform(poly_3.transform(X_train))))))

RMSE on training data for Standardized Elastic Regression with degree 3 :  721.1266235161161


In [98]:
print('RMSE on test data for Standardized Elastic Regression with degree 3 : ',np.sqrt(mean_squared_error(y_test, elastic_poly_3.predict(poly_scaler_3.transform(poly_3.transform(X_test))))))

RMSE on test data for Standardized Elastic Regression with degree 3 :  5215630.378085028


By adding L1/L2 regularization on polynomial regression of order 3, the model performance on test data improved by 7.6\*10^16 to 18681. The generalizability of the model has improved very highly. From the result of normal polynomial regression with degree 3 the training error was only 17 however the test error was in 10^16 which is a very high error, and that was showing that the model was overfitting. But by adding regularization the training error increased by a little however the test error reduced from factor of 10^16 to 10^4.
Eventhough, the model performance is not good, but the adding of penalty and its effect on model generalizibility can be clearly seen. 

#### Lastly, the addition of the regularization has improved the model generalazibility, plus improved the training time by alot. However, the model performance is not as expected, but impact of regularization can be clearly seen from the results above. The reason behind the bad model performance is because of the data. The NAN value filled with zeros is one of the reason why the model was not able to perform well. Since, introducing zeros introduces a noise in the data, which means the data behavior is completely changed due to it.